crates/valence_command/src/argument.rs

```diff
@@ -0,0 +1,1418 @@
+        let begin = reader.cursor();
+        let prop_value_str = reader.read_unquoted_str();
+        let prop_value = match PropValue::from_str(prop_value_str) {
+            Some(v) => v,
+            None => {
+                return ParsingResult {
+                    suggestions: Some((begin..reader.cursor(), BlockSuggestions::PropValue)),
+                    result: Err((
+                        begin..reader.cursor(),
+                        BlockError::PropInvalid {
+                            kind: kind_str,
+                            name: prop_name_str,
+                            value: prop_value_str,
+                        },
+                    )),
+                }
+            }
+        };
+
+        ParsingResult {
+            suggestions: None,
+            result: Ok(Some((prop_name, prop_value))),
+        }
+    }
+}
+
+impl<'a> BrigadierArgument<'a> for BlockPredicate<'a> {
+    fn parser(_data: Option<&Self::Data>) -> Parser<'a> {
+        Parser::BlockPredicate
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct BlockStateArgument {
+    pub state: BlockState,
+    pub tags: Option<Compound>,
+}
+
+impl<'a> Parsable<'a> for BlockStateArgument {
+    type Data = ();
+
+    type Error = BlockError<'a>;
+
+    type Suggestions = BlockSuggestions;
+
+    fn parse(
+        _data: Option<&Self::Data>,
+        reader: &mut StrReader<'a>,
+        purpose: ParsingPurpose,
+    ) -> ParsingResult<Self, Self::Suggestions, Self::Error> {
+        let begin = reader.cursor();
+        let kind_str = reader.read_ident_str();
+
+        let mut state = BlockState::from_kind(match BlockKind::from_str(kind_str) {
+            Some(kind) => kind,
+            None => {
+                return ParsingResult {
+                    suggestions: Some((begin..reader.cursor(), BlockSuggestions::BlockKind)),
+                    result: Err((begin..reader.cursor(), BlockError::Kind(kind_str))),
+                }
+            }
+        });
+
+        parsing_ret_err!(read_block_props(
+            reader,
+            purpose,
+            kind_str,
+            |prop_name, prop_value| {
+                state = state.set(prop_name, prop_value);
+                ParsingResult::ok()
+            }
+        ));
+
+        ParsingResult {
+            suggestions: None,
+            result: Ok(match purpose {
+                ParsingPurpose::Reading => Some(Self { state, tags: None }),
+                ParsingPurpose::Suggestion => None,
+            }),
+        }
+    }
+}
+
+impl<'a> BrigadierArgument<'a> for BlockStateArgument {
+    fn parser(_data: Option<&Self::Data>) -> Parser<'a> {
+        Parser::BlockState
+    }
+}
+
+/// Represent any enum that has no fields in its variants.
+///
+/// Brigadier's CEnums are:
+/// - color
+/// - dimension
+/// - entity_anchor
+/// - gamemode
+/// - heightmap
+/// - operation
+/// - swizzle (?)
+/// - template mirror
+/// - template rotation
+pub trait CEnum: Sized {
```

Comment:
What does the name `CEnum` mean? This isn't clear to me.

Reply:
C-Like enum. I understand that It is bad naming but I can not think of other one for now